tests/firecrackerdrivertest.c

```diff
@@ -0,0 +1,302 @@
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library;  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+
+#include <config.h>
+
+#include "testutils.h"
+
+#include "virlog.h"
+
+#ifdef WITH_FC
+
+#include "vircommand.h"
+
+# define VIR_FROM_THIS VIR_FROM_NONE
+
+VIR_LOG_INIT("tests.firecrackerdrivertest");
+
+static int
+testDriverConnection(const void *opaque G_GNUC_UNUSED)
+{
+    int ret = 0;
+
+    virConnectPtr conn = virConnectOpen("fc:///system");
+
+    if (conn == NULL)
+        ret = -1;
+
+    if (virConnectClose(conn) < 0)
+        ret = -1;
+
+    return ret;
+}
+
+static void
+downloadKernelAndRootFS(void)
+{
+    const char *arch = virArchToString(virArchFromHost());
+    g_autofree const char *kernel_url = g_strdup_printf(
+        "https://s3.amazonaws.com/spec.ccfc.min/img/quickstart_guide/%s/kernels/vmlinux.bin",
+        arch);
+    g_autofree const char *rootfs_url = g_strdup_printf(
+        "https://s3.amazonaws.com/spec.ccfc.min/img/quickstart_guide/%s/rootfs/bionic.rootfs.ext4",
+        arch);
+
+    g_autofree const char *download_kernel_path = g_strdup_printf(
+        "%s/firecrackerdriverdata/hello-vmlinux-test.bin",
+        abs_srcdir);
+    g_autofree const char *download_rootfs_path = g_strdup_printf(
+        "%s/firecrackerdriverdata/hello-rootfs-test.ext4",
+        abs_srcdir);
+
+    virCommand *cmd = virCommandNew("curl");
+
+    virCommandAddArg(cmd, "-fsSL");
+    virCommandAddArg(cmd, "-o");
+    virCommandAddArg(cmd, download_kernel_path);
+    virCommandAddArg(cmd, kernel_url);
+
+    if (virCommandRun(cmd, NULL) < 0) {
+        VIR_WARN("Error downloading test kernel from link: %s",
+                 kernel_url);
+        return;
+    }
+
+    virCommandFree(cmd);
+
+    cmd = virCommandNew("curl");
+
+    virCommandAddArg(cmd, "-fsSL");
+    virCommandAddArg(cmd, "-o");
+    virCommandAddArg(cmd, download_rootfs_path);
+    virCommandAddArg(cmd, rootfs_url);
+
+    if (virCommandRun(cmd, NULL) < 0) {
+        VIR_WARN("Error downloading test rootfs from link: %s",
+                 rootfs_url);
+        return;
+    }
+
+    VIR_DEBUG("Finished downloading");
+}
+
+static virConnectPtr test_conn;
+
+static int
+testInitiateConnection(const void *opaque G_GNUC_UNUSED)
+{
+    test_conn = virConnectOpen("fc:///system");
+
+    if (test_conn == NULL)
+        return -1;
+
+    return 0;
+}
+
+static int
+testDefineDomain(const void *opaque G_GNUC_UNUSED)
+{
+    const char *path = g_strdup_printf("%s/firecrackerdriverdata/%s.xml",
+                                      abs_srcdir,
+                                      "test_domain");
+
+    char *xml = NULL;
```

Comment:
this string isn't `free`'d

Reply:
Is it? where?

Reply:
I added `g_autofree` to the `path` variable. Now i see i should do this with the `xml` one too